pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+package crypto
+
+import (
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/ecdsa"
+	crand "crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	mrand "math/rand"
+	"strconv"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethCrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"github.com/ethersphere/swarm/log"
+)
+
+// We need a singleton here?
+var cryptoBackend defaultCryptoBackend
+var createPadding = true
+
+const (
+	aesNonceLength = 12 // in bytes; for more info please see cipher.gcmStandardNonceSize & aesgcm.NonceSize()
+	keyIDSize      = 32 // in bytes
+
+	flagsLength             = 1
+	payloadSizeFieldMaxSize = 4
+	signatureLength         = 65      // in bytes
+	padSizeLimit            = 256     // just an arbitrary number, could be changed without breaking the protocol
+	SizeMask                = byte(3) // mask used to extract the size of payload size field from the flags
+	signatureFlag           = byte(4)
+	aesKeyLength            = 32 // in bytes
+
+	defaultPaddingByteSize = 16
+)
+
+type WrapParams struct {
+	Src    *ecdsa.PrivateKey
+	Dst    *ecdsa.PublicKey
+	KeySym []byte
+}
+
+// Contains a successfully decrypted message prior to parsing and validating
+type ReceivedMessage interface {
```

Comment:
Why do we need to export this type? Can we make it so that the calling code in `pss` package only handles the (new in followup PR) `Message` type?

Reply:
This type encapsulates the result of a decryption: plaintext(Payload) and `PublicKey` signing it.
`PublicKey` is then used in `keystore` to obtain the `from PssAddress`.